src/parquet/arrow/reader.cc

```diff
@@ -1319,29 +1336,21 @@ Status PrimitiveImpl::NextBatch(int batch_size, std::shared_ptr<Array>* out) {
       TYPED_BATCH_CASE(BINARY, ::arrow::BinaryType, ByteArrayType)
       TYPED_BATCH_CASE(DATE32, ::arrow::Date32Type, Int32Type)
       TYPED_BATCH_CASE(DATE64, ::arrow::Date64Type, Int32Type)
-    case ::arrow::Type::FIXED_SIZE_BINARY: {
-      int32_t byte_width =
-          static_cast<::arrow::FixedSizeBinaryType*>(field_->type().get())->byte_width();
-      return ReadFLBABatch<::arrow::FixedSizeBinaryType>(batch_size, byte_width, out);
-      break;
-    }
+    case ::arrow::Type::FIXED_SIZE_BINARY:
+      return ReadFLBABatch<::arrow::FixedSizeBinaryType>(batch_size, type, out);
+    case ::arrow::Type::DECIMAL:
+      return ReadFLBABatch<::arrow::DecimalType>(batch_size, type, out);
     case ::arrow::Type::TIMESTAMP: {
-      ::arrow::TimestampType* timestamp_type =
-          static_cast<::arrow::TimestampType*>(field_->type().get());
+      const auto& timestamp_type = std::static_pointer_cast<::arrow::TimestampType>(type);
```

Comment:
As you already touch this line, change this to `static_cast<const ::arrow::TimestampType&>`. This should save us from some unnecessary reference counting.